### ISSUE TITLE
AsyncSemaphore not dequeuing a cancelled task

### ISSUE THREAD

**Issue by josh-green** [webappsuk/CoreLibraries#55]
A scenario has been found where a task can be en-queued that has already been cancelled. In the following code a task can be cancelled after the `IsCancellationRequested` check, but there is nothing setup to dequeue in the case of cancellation immediately after that.

https://github.com/webappsuk/CoreLibraries/blob/033037b3234a446753a0f846b4734ed7f91a9bd0/Utilities/Threading/AsyncSemaphore.cs#L134-L140

The result of this is that `waiter.Task.WithCancellation(token)` throws and the task is never dequeued. 
This could be solved by registering a cancellation action on the token:

```csharp
if (token.CanBeCanceled)
{
	token.Register(() =>
	{
		waiter.TrySetCanceled();
	});
}
```

Also to note that the current implementation of AsyncSemaphore in Core Libraries could be classed as superseded by the [SemaphoreSlim](https://docs.microsoft.com/en-us/dotnet/api/system.threading.semaphoreslim?view=netframework-4.7.1) .NET type. It doesn't have the disposable Waits that Core Libraries does, so we would just wrap the SemaphoreSlim and expose to get the same functionality.

**Comment by thargy** [comment]
I'm not sure the lock is required, I believe this could be rewritten without a lock, using code similar to:

```csharp

	/// <summary>
	/// Gets or sets the maximum count.
	/// </summary>
	/// <value>
	/// The maximum count.
	/// </value>
	/// <exception cref="ArgumentOutOfRangeException">value</exception>
	public int MaxCount
	{
		get => _maxCount;
		set
		{
			if (value < 1) throw new ArgumentOutOfRangeException(nameof(value));
			if (value == _maxCount) return;
			
			int maxCount;
			do {
				maxCount = _maxCount;
			} while (Interlocked.CompareExchange(ref _maxCount, value, maxCount) != maxCount);
			
			if (value > maxCount)
				_semaphore.Release(value-maxCount);
		}
	}

	/// <summary>
	/// Waits on the semaphore.
	/// </summary>
	/// <param name="token">The optional cancellation token.</param>
	/// <returns>Task.</returns>
	/// <remarks><para>This is best used with a <see langword="using"/> statement.</para></remarks>
	[NotNull]
	public async Task<IDisposable> WaitAsync(CancellationToken token = default(CancellationToken))
	{
		await _semaphore.WaitAsync(token).ConfigureAwait(false);
		Interlocked.Increment(ref _currentCount);
		return _releaser;
	}

	/// <summary>
	/// Releases any waiters waiting on the semaphore.
	/// </summary>
	public void Release()
	{
		int currentCount;
		do
		{
			currentCount = _currentCount;
			if (currentCount == 0) return;
		} while (Interlocked.CompareExchange(ref _currentCount, currentCount - 1, currentCount) != currentCount);

		if (currentCount <= _maxCount)
			_semaphore.Release();
	}
```

Though this would need careful consideration for race conditions - mostly when changing `MaxCount`.

If the `MaxCount` was readonly (no setter) then this would be trivial and the above code should be faster than using a lock.  We have to question how necessary changing the MaxCount after creation is, it seems like a use case that may never get used in practice, and yet could have a serious performance impact.

Can we evaluate the performance impact of using `Interlocked` instead of a lock.  As this is using `async` the overhead of that part of the framework may drown out the lock anyway, in which case there's little point in changing.

**Comment by josh-green** [comment]
Following some testing by @billings7 it was decided that the performance difference did not outweigh the potential risk of race conditions, as well as the reduction in maintainability.
